app/com/linkedin/drelephant/tuning/hbt/MRApplicationData.java

```diff
@@ -0,0 +1,375 @@
+      }
+      if (appHeuristicResultDetails.name.equals(MAX_PHYSICAL_MEMORY)) {
+        usedPhysicalMemoryMB = (double) MemoryFormatUtils.stringToBytes(appHeuristicResultDetails.value);
+      }
+      if (appHeuristicResultDetails.name.equals(MAX_TOTAL_COMMITTED_HEAP_USAGE)) {
+        usedHeapMemoryMB = (double) MemoryFormatUtils.stringToBytes(appHeuristicResultDetails.value);
+      }
+    }
+    addCounterData(new String[]{functionType + " " + MAX_VIRTUAL_MEMORY, functionType + " " + MAX_PHYSICAL_MEMORY,
+            functionType + " " + MAX_TOTAL_COMMITTED_HEAP_USAGE}, usedVirtualMemoryMB, usedPhysicalMemoryMB,
+        usedHeapMemoryMB);
+
+    logDebuggingStatement(" Used Physical Memory " + yarnAppHeuristicResult.yarnAppResult.id + "_" + functionType + " "
+            + usedPhysicalMemoryMB,
+        " Used Virtual Memory " + yarnAppHeuristicResult.yarnAppResult.id + "_" + functionType + " "
+            + usedVirtualMemoryMB,
+        " Used heap Memory " + yarnAppHeuristicResult.yarnAppResult.id + "_" + functionType + " " + usedHeapMemoryMB);
+
+    Double memoryMB = max(usedPhysicalMemoryMB, usedVirtualMemoryMB / (VIRTUALMEMORY_TO_PHYSICALMEMORY_RATIO));
+    Double heapSizeMax =
+        TuningHelper.getHeapSize(min(HEAPSIZE_TO_MAPPERMEMORY_SAFE_RATIO * memoryMB, usedHeapMemoryMB));
+    Double containerSize = TuningHelper.getContainerSize(memoryMB);
+    addParameterToSuggestedParameter(heapSizeMax, containerSize, yarnAppHeuristicResult.yarnAppResult.id, functionType);
+  }
+
+  private void addParameterToSuggestedParameter(Double heapSizeMax, Double containerSize, String id,
+      String functionType) {
+    if (functionType.equals(Mapper.name())) {
+      addMapperMemoryAndHeapToSuggestedParameter(heapSizeMax, containerSize, id);
+    } else {
+      addReducerMemoryAndHeapToSuggestedParameter(heapSizeMax, containerSize, id);
+    }
+  }
+
+  private void addMapperMemoryAndHeapToSuggestedParameter(Double heapSizeMax, Double containerSize,
+      String heuristicsResultID) {
+    suggestedParameter.put(MAPPER_MEMORY, containerSize);
+    suggestedParameter.put(MAPPER_HEAP, heapSizeMax);
+    logDebuggingStatement(" Memory Assigned " + heuristicsResultID + "_Mapper " + suggestedParameter.get(MAPPER_MEMORY),
+        " Heap Assigned " + heuristicsResultID + "_Mapper " + suggestedParameter.get(MAPPER_HEAP));
+  }
+
+  private void addReducerMemoryAndHeapToSuggestedParameter(Double heapSizeMax, Double containerSize,
+      String heuristicsResultID) {
+    suggestedParameter.put(REDUCER_MEMORY, containerSize);
+    suggestedParameter.put(REDUCER_HEAP, heapSizeMax);
+    logDebuggingStatement(
+        " Memory Assigned " + heuristicsResultID + "_Reducer " + suggestedParameter.get(REDUCER_MEMORY),
+        " Heap Assigned " + heuristicsResultID + "_Reducer " + suggestedParameter.get(REDUCER_HEAP));
+  }
+
+  private void processForNumberOfTask(AppHeuristicResult yarnAppHeuristicResult, String functionType) {
```

Comment:
Splitting this into two different function would help, one for map and one for reduce. There is nothing common here between map and reduce. It will avoid the if clauses as well.

Reply:
Have created two more methods processForNumberOfTaskMapper,processForNumberOfTaskReducer . But it would be better to keep processForNumberOfTask method which will call above two methods . Its better to have processForNumberOfTask as balnket methods , which internally calls above two methods .